Add MFEM point and line sampler

k-collie opened the pull request:
Closes #31249

## Design
- `MFEMValueSamplerBase` with shared functionality
    - composes `mfem::FindPointsGSLIB` to do interpolation
- `MFEMLineValueSampler` and `MFEMPointValueSampler` derived classes

## Impact
- adds extra build step for building MFEM
- adds new `vectorpostprocessors` for MFEM problems
- will enable future work to parallelise MFEM test output

## Notes

- probably needs a change to CI scripts to add `gslib` build step

k-collie committed:
Write MFEMVectorPostprocessor

refs #31249

k-collie committed:
Write MFEMValueSamplerBase

refs #31249

k-collie committed:
Write MFEMLineValueSampler

refs #31249

k-collie committed:
Write MFEMPointValueSampler

refs #31249

k-collie committed:
Write MFEMLineValueSampler tests

refs #31249

k-collie committed:
Write MFEMPointValueSampler tests

refs #31249

k-collie committed:
MFEMValueSamplerBase add check that all points were found

refs #31249

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3112648/), step Clang format on 04c8b0f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31250/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31250/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b4cb6f66255664d96e7cad21fd197653c70deeef`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3112648/), step Versioner verify on 04c8b0f wanted to post the following:

### Versioner templates

Found 14 templates, 0 failed


### Versioner influential files

Found 59 influential files, 2 changed, 1 added, 0 removed

| package   | status   | file                                |
|-----------|----------|-------------------------------------|
| moose-dev | CHANGE   | apptainer/moose-dev.def             |
| moose-dev | NEW      | scripts/update_and_rebuild_gslib.sh |
| moose-dev | CHANGE   | scripts/update_and_rebuild_mfem.sh  |

### Versioner versions

Found 9 packages, 1 changed, 1 failed

| package   | status    | hash    | to hash   | version    | to version   |
|-----------|-----------|---------|-----------|------------|--------------|
| moose-dev | NEED BUMP | 46da177 | f9c54b9   | 2025.07.22 | 2025.07.22   |

Verification failed.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3112701/), step Versioner verify on 374d8b8 wanted to post the following:

### Versioner templates

Found 14 templates, 0 failed


### Versioner influential files

Found 59 influential files, 2 changed, 1 added, 0 removed

| package   | status   | file                                |
|-----------|----------|-------------------------------------|
| moose-dev | CHANGE   | apptainer/moose-dev.def             |
| moose-dev | NEW      | scripts/update_and_rebuild_gslib.sh |
| moose-dev | CHANGE   | scripts/update_and_rebuild_mfem.sh  |

### Versioner versions

Found 9 packages, 1 changed, 1 failed

| package   | status    | hash    | to hash   | version    | to version   |
|-----------|-----------|---------|-----------|------------|--------------|
| moose-dev | NEED BUMP | 46da177 | f9c54b9   | 2025.07.22 | 2025.07.22   |

Verification failed.

k-collie committed:
Add docs for new MFEMVectorPostprocessor classes

refs #31249

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3123332/), step Versioner verify on f01253a wanted to post the following:

### Versioner templates

Found 14 templates, 0 failed


### Versioner influential files

Found 59 influential files, 2 changed, 1 added, 0 removed

| package   | status   | file                                |
|-----------|----------|-------------------------------------|
| moose-dev | CHANGE   | apptainer/moose-dev.def             |
| moose-dev | NEW      | scripts/update_and_rebuild_gslib.sh |
| moose-dev | CHANGE   | scripts/update_and_rebuild_mfem.sh  |

### Versioner versions

Found 9 packages, 1 changed, 1 failed

| package   | status    | hash    | to hash   | version    | to version   |
|-----------|-----------|---------|-----------|------------|--------------|
| moose-dev | NEED BUMP | 46da177 | f9c54b9   | 2025.07.22 | 2025.07.22   |

Verification failed.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3123598/), step Versioner verify on 5750238 wanted to post the following:

### Versioner templates

Found 14 templates, 0 failed


### Versioner influential files

Found 59 influential files, 3 changed, 1 added, 0 removed

| package   | status   | file                                |
|-----------|----------|-------------------------------------|
| moose-dev | CHANGE   | apptainer/moose-dev.def             |
| moose-dev | CHANGE   | conda/moose-dev/meta.yaml           |
| moose-dev | NEW      | scripts/update_and_rebuild_gslib.sh |
| moose-dev | CHANGE   | scripts/update_and_rebuild_mfem.sh  |

### Versioner versions

Found 9 packages, 1 changed, 0 failed

| package   | status   | hash    | to hash   | version    | to version   |
|-----------|----------|---------|-----------|------------|--------------|
| moose-dev | CHANGE   | 46da177 | e892344   | 2025.07.22 | 2025.08.12   |

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3130057/), step Versioner verify on 39f19c0 wanted to post the following:

### Versioner templates

Found 14 templates, 0 failed


### Versioner influential files

Found 59 influential files, 3 changed, 1 added, 0 removed

| package   | status   | file                                |
|-----------|----------|-------------------------------------|
| moose-dev | CHANGE   | apptainer/moose-dev.def             |
| moose-dev | CHANGE   | conda/moose-dev/meta.yaml           |
| moose-dev | NEW      | scripts/update_and_rebuild_gslib.sh |
| moose-dev | CHANGE   | scripts/update_and_rebuild_mfem.sh  |

### Versioner versions

Found 9 packages, 1 changed, 0 failed

| package   | status   | hash    | to hash   | version    | to version   |
|-----------|----------|---------|-----------|------------|--------------|
| moose-dev | CHANGE   | 7cad19c | d2e623c   | 2025.08.06 | 2025.08.12   |

k-collie commented:
@nmnobre told me about [this PR](https://github.com/mfem/mfem/pull/5001) on MFEM which we could use instead of this handling of the GSLIB dependency.

k-collie committed:
Fix mesh 2D meshes for MFEMValueSamplerBase

refs #31249

Before MFEM vector was hard-coded with LIBMESH_DIM spatial dimensions,
now the dimensions are read from the MFEM mesh class. This avoids a
segfault when writing to declared points (which have dimensions set by
mesh space dimension).

k-collie commented:
Re-based on next after `gslib` change merged and added fix for sampling from 2D meshes.

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3540942/), step Results summary on f9ccfa9 wanted to post the following:

## Framework test summary

Compared against 4bccd3a in job [civet.inl.gov/job/3537990](https://civet.inl.gov/job/3537990).

### Added tests

| Test                                                                           | Time (s)   | Memory (MB)   |
|--------------------------------------------------------------------------------|------------|---------------|
| `mfem/vectorpostprocessors/point_value_sampler.MFEMPointValueSamplerDiffusion` | 1.04       | 0.00          |
| `mfem/vectorpostprocessors/line_value_sampler.MFEMLineValueSamplerDiffusion`   | 1.00       | 0.00          |
| `mfem/vectorpostprocessors/line_value_sampler.MFEMLineValueSamplerCurlCurl`    | 0.82       | 0.00          |
| `mfem/vectorpostprocessors/point_value_sampler.MFEMPointValueSamplerCurlCurl`  | 0.80       | 0.00          |
## Modules test summary

Compared against 4bccd3a in job [civet.inl.gov/job/3537990](https://civet.inl.gov/job/3537990).

No change

alexanderianblair commented on framework/doc/content/source/mfem/vectorpostprocessors/MFEMLineValueSampler.md:
```diff
@@ -0,0 +1,23 @@
+## Overview
+
+This class performs accurate interpolation of scalar and vector quantities
+along a specified line using MFEM's `FindPointsGSLIB` `gslib` integration.
```
```suggestion
along a specified line using MFEM's `FindPointsGSLIB` `gslib` interpolation.
```

alexanderianblair commented on framework/doc/content/source/mfem/vectorpostprocessors/MFEMPointValueSampler.md:
```diff
@@ -0,0 +1,23 @@
+## Overview
+
+This class performs accurate interpolation of scalar and vector quantities at a
+set of specified points using MFEM's `FindPointsGSLIB` `gslib` integration.
```
```suggestion
set of specified points using MFEM's `FindPointsGSLIB` `gslib` interpolation.
```

alexanderianblair commented on test/tests/mfem/vectorpostprocessors/line_value_sampler/line_value_sampler_curlcurl.i:
```diff
@@ -0,0 +1,129 @@
+[Mesh]
+  type = MFEMMesh
+  file = ../../mesh/small_fichera.mesh
+  dim = 3
```
```suggestion
```

k-collie committed:
Apply suggestions from Alex B

refs #31250

Co-authored-by: Alexander Blair <35571481+alexanderianblair@users.noreply.github.com>

k-collie committed:
Wrap text in comment

refs #31250

k-collie committed:
Remove redundant dim=3 specification from tests

refs #31250

k-collie committed:
Run MFEM sampler tests on CPU and GPU

refs #31250

k-collie committed:
Add link to MFEM GSLib how to

refs #31250

k-collie committed:
Remove redundant syntax from tests

refs #31250

k-collie committed:
Add design and issue fields for MFEM interp tests

refs #31250

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3551435/), step Clang format on 0171227 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31250/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31250/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e4407b31a57ef5bd7bcb870595bbe7897b813978`

k-collie committed:
Switch to lowerCamelCase for functions

refs #31250

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3553302/), step Results summary on dd56f76 wanted to post the following:

## Framework test summary

Compared against 4b9b580 in job [civet.inl.gov/job/3550358](https://civet.inl.gov/job/3550358).

### Added tests

| Test                                                                           | Time (s)   | Memory (MB)   |
|--------------------------------------------------------------------------------|------------|---------------|
| `mfem/vectorpostprocessors/point_value_sampler.MFEMPointValueSamplerDiffusion` | 0.95       | 0.00          |
| `mfem/vectorpostprocessors/line_value_sampler.MFEMLineValueSamplerDiffusion`   | 0.95       | 0.00          |
| `mfem/vectorpostprocessors/point_value_sampler.MFEMPointValueSamplerCurlCurl`  | 0.77       | 0.00          |
| `mfem/vectorpostprocessors/line_value_sampler.MFEMLineValueSamplerCurlCurl`    | 0.76       | 0.00          |
## Modules test summary

Compared against 4b9b580 in job [civet.inl.gov/job/3550358](https://civet.inl.gov/job/3550358).

No change
